src/contract.ts

```diff
@@ -1,376 +1,479 @@
+        if (!payer) throw new Error('MISSING_PAYER_ACCOUNT'); // @FIXME: add error types
+
+        const {
+            accounts = [],
+            writableAccounts = [],
+            programDerivedAddresses = [],
+            signers = [],
+            sender = payer.publicKey,
+            value = 0,
+            simulate = false,
+            confirmOptions = {
+                commitment: 'confirmed',
+                skipPreflight: false,
+                preflightCommitment: 'processed',
+            },
+        } = options ?? {};
+
+        const hash = keccak256(Buffer.from(name));
+        const seeds = programDerivedAddresses.map((pda) => pda.seed);
+        const input = this.interface.encodeDeploy(constructorArgs);
+
+        const data = Buffer.concat([
+            this.storage.toBuffer(), //                     storage @FIXME: these comments are kind of useless
+            sender.toBuffer(), //                           sender  @FIXME: better to explain why, not what
+            Buffer.from(numToPaddedHex(value), 'hex'), //   value
+            Buffer.from(hash.substr(2, 8), 'hex'), //       hash
+            encodeSeeds(seeds), //                          seeds
+            Buffer.from(input.replace('0x', ''), 'hex'), // input
+        ]);
+
+        // @FIXME: why are so many of these keys commented out?
+        const keys = [
+            // @FIXME: should all these PDAs really be writable?
+            // ...programDerivedAddresses.map((pubkey) => ({
+            //   pubkey,
+            //   isSigner: false,
+            //   isWritable: true,
+            // })),
+            {
+                pubkey: storage.publicKey,
+                isSigner: false,
+                isWritable: true,
+            },
+            // {
+            //   pubkey: SYSVAR_CLOCK_PUBKEY,
+            //   isSigner: false,
+            //   isWritable: false,
+            // },
+            // {
+            //   pubkey: PublicKey.default,
+            //   isSigner: false,
+            //   isWritable: false,
+            // },
+            ...accounts.map((pubkey) => ({
+                pubkey,
+                isSigner: false,
+                isWritable: false,
+            })),
+            ...writableAccounts.map((pubkey) => ({
+                pubkey,
+                isSigner: false,
+                isWritable: true,
+            })),
+        ];
+
+        const lamports = await this.connection.getMinimumBalanceForRentExemption(space, confirmOptions.commitment);
+
+        const transaction = new Transaction().add(
+            SystemProgram.createAccount({
+                fromPubkey: payer.publicKey,
+                newAccountPubkey: storage.publicKey,
+                lamports,
+                space,
+                programId: this.program,
+            }),
+            new TransactionInstruction({
+                keys,
+                programId: this.program,
+                data,
+            })
```

Comment:
We now create the storage account atomically in the same transaction, which should also allow simulation of deploys to work without creating the account, and will also avoid creating orphaned accounts if deploys fail.